packages/runtime/container-runtime/src/opDecompressor.ts

```diff
@@ -0,0 +1,80 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { decompress } from "lz4js";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { assert, IsoBuffer, Uint8ArrayToString } from "@fluidframework/common-utils";
+import { CompressionAlgorithms } from ".";
+
+/**
+ * State machine that "unrolls" contents of compressed batches of ops after decompressing them.
+ * This class relies on some implicit contracts defined below:
+ * 1. A compressed batch will have its first message with batch metadata set to true and compressed set to true
+ * 2. Messages in the middle of a compressed batch will have neither batch metadata nor the compression property set
+ * 3. The final message of a batch will have batch metadata set to false
+ * 4. An individually compressed op will have undefined batch metadata and compression set to true
+ */
+export class OpDecompressor {
+    private activeBatch = false;
+    private rootMessageContents: any | undefined;
+    private processedCount = 0;
+
+    constructor() {
+
+    }
+
+    public processMessage(message: ISequencedDocumentMessage): ISequencedDocumentMessage {
+        // Beginning of a compressed batch
+        // We're checking for compression = true or top level compression property so
+        // that we can enable compression without waiting on all ordering services
+        // to pick up protocol change. Eventually only the top level property should
+        // be used.
+        if (message.metadata?.batch === true
+            && (message.metadata?.compressed || (message as any).compression !== undefined)) {
+            assert(this.activeBatch === false, "shouldn't have multiple active batches");
+            if ((message as any).compression) {
```

Comment:
shouldn't we assert regardless of this being defined or not? I mean if it's compressed but we didn't specify the compression algorithm, doesn't it violate the contract?

Reply:
Originally we wanted to use the metadata field to avoid relying on servers to round-trip the `compression` property added to `IDocumentMessage` and `ISequencedDocumentMessage` so that this could be used experimentally soon. So in that case `compressed: true` in metadata without the `compression` field is a valid state that implies lz4 compression. Depending on what we want to do I can remove this state as you suggested and rip out all the metadata related code.

Reply:
thanks for clarifying